src/features/instantiate/useInstantiation.tsx

```diff
@@ -18,48 +21,58 @@ export interface Instantiation {
   prefill: Prefill;
 }
 
-function useInstantiateMutation() {
+function useInstantiateMutation(isUserInitiated: boolean) {
   const { doInstantiate } = useAppMutations();
   const navigate = useNavigate();
   const queryClient = useQueryClient();
   const currentLanguage = useCurrentLanguage();
+  const setNavigationEffect = useSetNavigationEffect();
 
   return useMutation({
     mutationKey: ['instantiate', 'simple'],
     mutationFn: (instanceOwnerPartyId: number) => doInstantiate(instanceOwnerPartyId, currentLanguage),
     onError: (error: HttpClientError) => {
       window.logError('Instantiation failed:\n', error);
     },
-    onSuccess: (data) => {
-      navigate(`/instance/${data.id}`);
-      queryClient.invalidateQueries({ queryKey: ['fetchApplicationMetadata'] });
+    onSuccess: async (data) => {
+      const targetLocation = `/instance/${data.id}`;
+      if (isUserInitiated) {
+        setNavigationEffect({ targetLocation, matchStart: true, callback: focusMainContent });
+      }
+      navigate(targetLocation);
+      await queryClient.invalidateQueries({ queryKey: ['fetchApplicationMetadata'] });
     },
   });
 }
 
-function useInstantiateWithPrefillMutation() {
+function useInstantiateWithPrefillMutation(isUserInitiated: boolean) {
   const { doInstantiateWithPrefill } = useAppMutations();
   const navigate = useNavigate();
   const queryClient = useQueryClient();
   const currentLanguage = useCurrentLanguage();
+  const setNavigationEffect = useSetNavigationEffect();
 
   return useMutation({
     mutationKey: ['instantiate', 'withPrefill'],
     mutationFn: (instantiation: Instantiation) => doInstantiateWithPrefill(instantiation, currentLanguage),
     onError: (error: HttpClientError) => {
       window.logError('Instantiation with prefill failed:\n', error);
     },
-    onSuccess: (data) => {
-      navigate(`/instance/${data.id}`);
-      queryClient.invalidateQueries({ queryKey: ['fetchApplicationMetadata'] });
+    onSuccess: async (data) => {
+      const targetLocation = `/instance/${data.id}`;
+      if (isUserInitiated) {
+        setNavigationEffect({ targetLocation, matchStart: true, callback: focusMainContent });
+      }
+      navigate(targetLocation);
+      await queryClient.invalidateQueries({ queryKey: ['fetchApplicationMetadata'] });
     },
   });
 }
 
-export const useInstantiation = () => {
+export const useInstantiation = (isUserInitiated: boolean) => {
```

Comment:
Umulig å skjønne hva dette argumentet betyr når man leser `useInstantiation(false/true)` uten å hoppe ut av nåværende kontekts for å sjekke 😅 hvis man ikke gjør det kan man fort nesten tro at det betyr "instansier"/"ikke instansier" for true/false respektivt, hehe. Kan du gjøre det om til et objekt?

Reply:
Siden dette forøvrig styrer en effekt som bare blir satt to steder, ville jeg heller droppet argumentet her og eksponert hele `mutation`, objektet. Da kan man sette en `onSuccess` callback når man faktisk kjører `mutate`-funksjonen. 
```
mutate(variables, {
  onError,
  onSettled,
  onSuccess,
})
``` 

https://tanstack.com/query/latest/docs/framework/react/reference/useMutation

Litt ryddigere imo og da slipper alle steder som bruker `useInstantiation` å forholde seg til noe som styrer noe man kanskje helst vil sette litt eksplisitt 😊

Reply:
Hmm, blir vel vanskeligere å opprettholde sjekken på `!hasAlreadyInstantiated` da, så jeg tenker jeg heller åpner opp for å sende inn `MutateOptions` :+1: